horovod/common/half.cc

```diff
@@ -0,0 +1,78 @@
+// Copyright 2018 Uber Technologies, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+
+#if __AVX__ && __F16C__
+#include <cpuid.h>
+#include <immintrin.h>
+#endif
+
+#include "half.h"
+
+namespace horovod {
+namespace common {
+
+#if __AVX__ && __F16C__
+// Query CPUID to determine AVX and F16C runtime support.
+bool is_avx_and_f16c() {
+  static bool initialized = false;
+  static bool result = false;
+  if (!initialized) {
+    unsigned int eax, ebx, ecx, edx;
+    if (__get_cpuid(1, &eax, &ebx, &ecx, &edx)) {
+      result = (ecx & bit_AVX) && (ecx & bit_F16C);
+    }
+    initialized = true;
+  }
+  return result;
+}
+#endif
+
+// float16 custom data type summation operation.
+void float16_sum(void* invec, void* inoutvec, int* len,
+                 MPI_Datatype* datatype) {
+  // cast invec and inoutvec to your float16 type
+  auto* in = (unsigned short*)invec;
+  auto* inout = (unsigned short*)inoutvec;
+
+  int i = 0;
```

Comment:
Why not initialize `i` within the for loop?

Reply:
Because if we have AVX, we will process all the blocks divisible by 8, and the rest will be processed as "leftovers" using the slow algorithm.  If we don't have AVX, `i` will start at 0.